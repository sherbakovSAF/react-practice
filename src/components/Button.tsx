import type { ButtonHTMLAttributes, PropsWithChildren } from "react";
import styles from "./Button.module.scss";
import clsx from "clsx";
interface ButtonProps
  extends PropsWithChildren<ButtonHTMLAttributes<HTMLButtonElement>> {
  view?: "primary" | "secondary";
}

const Button: React.FC<ButtonProps> = ({
  children,
  className,
  view = "primary",
  ...props
}) => {
  return (
    <button
      {...props}
      className={clsx(
        className,
        styles.btn,
        view === "secondary" && styles["btn--secondary"]
      )}
    >
      {children}
    </button>
  );
};

export default Button;
