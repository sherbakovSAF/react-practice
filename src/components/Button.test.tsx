import { test, expect, describe, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import Button from "./Button";

describe("Button", () => {
  test("has type='button' by default", () => {
    render(<Button>Кнопка</Button>);
    expect(screen.getByRole("button")).toHaveAttribute("type", "button");
  });
  test("accepts custom type", () => {
    render(<Button type="submit">Отправить</Button>);
    expect(screen.getByRole("button")).toHaveAttribute("type", "submit");
  });
  test("has base style class", () => {
    render(<Button>Стиль</Button>);
    expect(screen.getByRole("button").className).toMatch(/btn/);
  });
  test("adds secondary class when view='secondary'", () => {
    render(<Button view="secondary">Кнопка</Button>);
    expect(screen.getByRole("button").className).toMatch(/btn--secondary/);
  });
  test("merges passed className", () => {
    render(<Button className="custom-class">Кнопка</Button>);
    expect(screen.getByRole("button").className).toMatch(/custom-class/);
  });
  test("calls onClick handler", () => {
    const handleClick = vi.fn();
    render(<Button onClick={handleClick}>Клик</Button>);
    fireEvent.click(screen.getByRole("button"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
  test("supports disabled prop", () => {
    render(<Button disabled>Disabled</Button>);
    expect(screen.getByRole("button")).toBeDisabled();
  });

  test("supports aria-label", () => {
    render(<Button aria-label="Моя кнопка" />);
    expect(screen.getByLabelText("Моя кнопка")).toBeInTheDocument();
  });
});
