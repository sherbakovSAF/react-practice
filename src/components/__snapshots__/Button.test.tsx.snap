// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders default button snapshot 1`] = `
<DocumentFragment>
  <button
    class="_btn_d3d534"
    type="button"
  >
    Кнопка
  </button>
</DocumentFragment>
`;

exports[`Button > renders secondary button snapshot 1`] = `
<DocumentFragment>
  <button
    class="_btn_d3d534 _btn--secondary_d3d534"
    type="button"
  >
    Кнопка
  </button>
</DocumentFragment>
`;

exports[`Button > renders with custom class snapshot 1`] = `
<DocumentFragment>
  <button
    class="custom-class _btn_d3d534"
    type="button"
  >
    Кнопка
  </button>
</DocumentFragment>
`;
